src/processing/processing_subtask_queue_manager.cpp

```diff
@@ -95,89 +107,173 @@ namespace sgns::processing
+                    }
+                }
+                // If we're the owner, update the processed IDs in the queue from our local set
+                queue->mutable_processing_queue()->clear_processed_subtask_ids();
+                for (const auto& processedId : m_processedSubTaskIds) {
+                    queue->mutable_processing_queue()->add_processed_subtask_ids(processedId);
+                }
 
-            // Map subtask IDs to subtask indices
-            std::vector<int> unprocessedSubTaskIndices;
-            for ( int subTaskIdx = 0; subTaskIdx < queue->subtasks().items_size(); ++subTaskIdx )
-            {
-                const auto &subTaskId = queue->subtasks().items( subTaskIdx ).subtaskid();
-                if ( m_processedSubTaskIds.find( subTaskId ) == m_processedSubTaskIds.end() )
+                if (queueChanged)
                 {
-                    unprocessedSubTaskIndices.push_back( subTaskIdx );
+                    m_queue_timestamp_ = queue->processing_queue().last_update_timestamp();
+                    UpdateQueueTimestamp();
+                    m_logger->debug("QUEUE_PUBLISH_SUBTASK_PROCESSED_UPDATE: on {} at {}ms", m_localNodeId, m_queue_timestamp_);
+                    LogQueue();
+                    PublishSubTaskQueue();
                 }
             }
 
-            if ( m_processingQueue.UpdateQueue( queue->mutable_processing_queue(), unprocessedSubTaskIndices ) )
+            // Now map subtask IDs to subtask indices based on the updated m_processedSubTaskIds
+            std::vector<int> unprocessedSubTaskIndices;
+            UpdateUnprocessedSubTaskIndices(queue.get(), unprocessedSubTaskIndices);
+
+            if (m_processingQueue.UpdateQueue(queue->mutable_processing_queue(), unprocessedSubTaskIndices))
             {
-                m_queue.swap( queue );
+                m_logger->debug("QUEUE_LOCAL_UPDATE: on {} at {}ms", m_localNodeId, m_queue_timestamp_);
+                m_queue.swap(queue);
                 LogQueue();
-                return true;
+                queueChanged = true;
             }
         }
-        return false;
+        return queueChanged;
     }
 
     void ProcessingSubTaskQueueManager::ProcessPendingSubTaskGrabbing()
     {
-        // The method has to be called in scoped lock of queue mutex
         m_dltGrabSubTaskTimeout.expires_at( boost::posix_time::pos_infin );
 
+        m_logger->trace("QUEUE_PROCESS_PENDING: for node {} at {}ms.", m_localNodeId, m_queue_timestamp_);
+
         // Update queue timestamp based on current ownership duration
         UpdateQueueTimestamp();
 
-        while ( !m_onSubTaskGrabbedCallbacks.empty() )
+        CheckActiveCount();
+
+        bool losingOwnership = false;
+        bool lockReleased = false;
+
+        while ( !losingOwnership &&
+                !m_onSubTaskGrabbedCallbacks.empty() &&
+                m_processedSubtasksInCurrentOwnership < m_maxSubtasksPerOwnership )
```

Comment:
I'd add a parenthesis 
```suggestion
                (m_processedSubtasksInCurrentOwnership < m_maxSubtasksPerOwnership) )
```